ministryofjustice/laa-apply-for-criminal-legal-aid
# [CRIMAP-544] List uploaded evidence for in progress application

## Description of change
This PR introduces the initial functionality to upload evidence for a draft application.

The `DocumentController` receives the file via the submit form and creates a local `Document` record with attributes like file_name etc. One attribute will be nil at this point, `object_key`.

The document is validated on size and content criteria and if valid, the `Upload` service requests an upload presigned URL via the datastore API and the presigned URL is then used to perform the upload directly to the bucket.

Upon successful completion of the upload, the returned `object_key` is written back to the `Document` record, and the transaction is finalised.

Upload failures are handled in the controller for JS and non-JS versions.

## Link to relevant ticket

[CRIMAP-544](https://dsdmoj.atlassian.net/browse/CRIMAP-544)

## Notes for reviewer
Apologies for the chunky PR!

Note: we have disabled multi file uploads for the non js version as the request payload included all the documents and it was not uploading one document at a time. As it's very unlikely that anyone will use the non js version, we did not feel it was worth the technical effort to handle both transaction types so the upload has been limited to one file at a time. 

Note: Application types list has been lifted from another application and may need dev review but should do for the time being. Design and content need reviewing, particularly for the error text.

Instructions to test are below.

## Screenshots of changes (if applicable)

### Before changes:
<img width="1130" alt="Screenshot 2023-09-12 at 17 36 54" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/ead39be0-38da-4d5e-9622-36e56d2c9998">


### After changes:
<img width="1130" alt="Screenshot 2023-09-12 at 17 18 34" src="https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/8f59cf6e-7def-435f-95e6-76e7a0f895e7">

https://github.com/ministryofjustice/laa-apply-for-criminal-legal-aid/assets/51047911/c9c850a9-cdf8-4047-956a-fc9600b22e9d

## How to manually test the feature
This requires having both datastore and apply running

1. In your datastore repo, navigate to `main` and run `bundle install` to ensure gems are up to date. 
2. In one terminal, run `rails server` to spin up datastore 
3. In another terminal, run `docker-compose up localstack` to spin up localstack (required to upload to s3) 
4. In apply, navigate to this branch and run `bundle install` to ensure gems are up to date and spin up apply by running `rails server`. You may need to run a migration.
5. Then navigate to apply in your chosen browser and create a draft application (or use an existing draft application) and navigate to `/applications/{application_uuid}/steps/evidence/upload`. You should then be able to upload a file

Please disable JS in your chosen browser and test the non js upload functionality. 


[CRIMAP-544]: https://dsdmoj.atlassian.net/browse/CRIMAP-544?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
